overlays/bootstrap.nix

```diff
@@ -365,41 +377,33 @@ in {
             '' + installDeps targetPrefix);
         })));
 
-    # Need to use something from 8.8.x as the default to build aarch64 native compiler:
-    defaultCompilerNixName =
-        if final.targetPlatform.isAarch64 && final.buildPlatform.isAarch64
-        then "ghc883"
-        else "ghc865";
-    ghc = final.haskell-nix.compiler."${final.haskell-nix.defaultCompilerNixName}";
     # Both `cabal-install` and `nix-tools` are needed for `cabalProject`
     # to check materialized results.  We need to take care that when
     # it is doing this we do not check the materialization of the
     # tools used or there will be infinite recursion.
     # always has `checkMaterialization = false` to avoid infinite
     # recursion.
-    cabal-install-tool = args:
-      # Hackage version of cabal-install does not compiler with ghc 8.10.1 yet.
-      let compiler-nix-name = args.compiler-nix-name or
-            (if final.haskell-nix.defaultCompilerNixName == "ghc8101"
-              then "ghc883"
-              else final.haskell-nix.defaultCompilerNixName);
-      in final.haskell-nix.tool "cabal" ({
+    cabal-install-tool = {compiler-nix-name, ...}@args:
+      (final.haskell-nix.hackage-package ({
+        name = "cabal-install";
         version = "3.2.0.0";
         index-state = final.haskell-nix.internalHackageIndexState;
-        inherit compiler-nix-name;
         # When building cabal-install (only happens when checking materialization)
         # disable checking of the tools used to avoid infinite recursion.
-        cabal-install = final.evalPackages.haskell-nix.cabal-install-tool
-          (args // { checkMaterialization = false; });
-        nix-tools = final.evalPackages.haskell-nix.nix-tools-set
-          (args // { checkMaterialization = false; });
+        cabal-install = final.evalPackages.haskell-nix.cabal-install-unchecked.${compiler-nix-name};
+        nix-tools = final.evalPackages.haskell-nix.nix-tools-unchecked.${compiler-nix-name};
         materialized = ../materialized + "/${compiler-nix-name}/cabal-install";
-      } // args);
-    cabal-install = final.buildPackages.haskell-nix.cabal-install-tool {};
-    nix-tools-set = args:
+      } // args)).components.exes.cabal;
+    cabal-install = final.lib.mapAttrs (compiler-nix-name: _:
+      final.haskell-nix.cabal-install-tool { inherit compiler-nix-name; }) final.haskell-nix.compiler;
+    cabal-install-unchecked = final.lib.mapAttrs (compiler-nix-name: _:
+      final.haskell-nix.cabal-install-tool { inherit compiler-nix-name; checkMaterialization = false; }) final.haskell-nix.compiler;
+    # Use this where we still have not good way to choose GHC version
+    internal-cabal-install = final.haskell-nix.cabal-install.ghc883;
```

Comment:
Is it important that we use the same version for `internal-cabal-install` and `internal-nix-tools`? Should we pull out `internalToolsCompilerVersion`? Also this could do with a note saying when/if it's safe to update the compiler version used here. (TBH, might be overkill but I think maybe we should annotate every reference to a specific compiler version with a note saying when/if we can change it!)